### PR TITLE
Get rid of VARINT default argument

### PR DESCRIPTION
**Pull request by sipa**
This removes the need for the non-strandard use of variadic macros.

### CODE REVIEW

**Comment by fanquake** [comment]
Thanks. Concept ACK.

**Comment by MarcoFalke** [comment]
What is the risk or downside of using GNU extensions? Is this a stylistic issue?

**Comment by Empact** [comment]
Given the two current formats are:
* DEFAULT -> unsigned
* NONNEGATIVE_SIGNED -> signed

Assuming other modes aren't planned. How about using two separate macros to present them: `VARINT` and `VARINT_U` (or similar)?

Incidentally, I don't see anywhere we're currently checking that the NONNEGATIVE_SIGNED values are in fact nonnegative.

**Comment by sipa** [comment]
> What is the risk or downside of using GNU extensions? Is this a stylistic issue?

Ping @fanquake, @theuni.

> Assuming other modes aren't planned. How about using two separate macros to present them: `VARINT` and `VARINT_U` (or similar)?

@ryanofsky What do you think of this suggestion?

> Incidentally, I don't see anywhere we're currently checking that the NONNEGATIVE_SIGNED values are in fact nonnegative.

Good point. They're only used in local disk formats, but perhaps an assert makes sense.

**Comment by fanquake** [comment]
> Is this a stylistic issue?

@MarcoFalke No. The better place for this discussion would be #18088.

**Comment by laanwj** [comment]
> What is the risk or downside of using GNU extensions? Is this a stylistic issue?

It means that a compiler "merely" implementing the C++11 standard could not compile our code. 

Maybe all current C++ compilers implement these extensions at the moment, but how would we feel if this would be some proprietary MSVC extension used?

IMO, unless there is another strong reason that gives a clear advantage (say, in efficiency, or correctness), it's preferable to not use special extensions.

**Comment by MarcoFalke** [comment]
Concept ACK per @laanwj 

Slightly prefer the version by @Empact because it feels odd to explicitly pass in a value that is just called "default"

**Comment by practicalswift** [comment]
Concept ACK: inside the standard is obviously better than outside the standard :)

Talking about standards: people interested in this PR might be interested in reviewing #17208 ("Make all tests pass UBSan without using any UBSan suppressions") which needs review :)

**Comment by practicalswift** [comment]
ACK c2dc9811a05f86534bd814b533271fac6b7ac68e -- patch looks correct

File: src/serialize.h [hunk 1]
```diff
@@ -490,12 +490,12 @@ class Wrapper
 template<typename Formatter, typename T>
 static inline Wrapper<Formatter, T&> Using(T&& t) { return Wrapper<Formatter, T&>(t); }
 
-#define VARINT(obj, ...) Using<VarIntFormatter<__VA_ARGS__>>(obj)
+#define VARINT(obj, mode) Using<VarIntFormatter<mode>>(obj)
```
Comment (ryanofsky) [inline src/serialize.h]: Would suggest something like:

```c++
#define VARINT(obj)            VARINT_MODE(obj, VarIntMode::DEFAULT)
#define VARINT_MODE(obj, mode) Using<VarIntFormatter<mode>>(obj)
```

to reduce verbosity and number of changes needed here

**Review by ryanofsky — APPROVED** [review]
Code review ACK c2dc9811a05f86534bd814b533271fac6b7ac68e. Verified changes with

```sh
git checkout origin/pull/18087/head
git grep -l VARINT | xargs sed -i 's/VARINT(\([^,]\+\), VarIntMode::DEFAULT)/VARINT(\1\)/'
git diff HEAD^
```

**Comment by sipa** [comment]
Rebased, and switched to @ryanofsky's VARINT/VARINT_MODE approach (smaller diff!).

File: src/serialize.h (hunk 1 above)
Comment (sipa) [reply src/serialize.h]: Done.

File: src/serialize.h [hunk 2]
```diff
@@ -211,7 +211,7 @@ template<typename X> const X& ReadWriteAsHelper(const X& x) { return x; }
  *
  * Example use:
  *   struct FooFormatter {
- *     FORMATTER_METHODS(Class, obj) { READWRITE(obj.val1, VARINT(obj.val2)); }
+ *     FORMATTER_METHODS(Class, obj) { READWRITE(obj.val1, VARINT(obj.val2, VarIntMode::DEFAULT)); }
```
Comment (MarcoFalke) [inline src/serialize.h]: :eyes:

**Review by ryanofsky — APPROVED** [review]
Code review ACK ace3c7157e4038c96e94f8b66490ba62439967d4. Changes since last review: rebase and adding VARINT_MODE

File: src/serialize.h (hunk 2 above)
Comment (sipa) [reply src/serialize.h]: Aha, someone caught it!

(*clearly* I left that in to see if reviewers were paying attention)

Comment (sipa) [reply src/serialize.h]: Fixed.

**Review by ryanofsky — APPROVED** [review]
Code review ACK 0e0fa27acb74b4f0075afcf59a0dff51a21baddb. Only change since last review reverting outdated documentation change from earlier version of pr

**Comment by MarcoFalke** [comment]
ACK 0e0fa27acb74b4f0075afcf59a0dff51a21baddb 📯

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 0e0fa27acb74b4f0075afcf59a0dff51a21baddb 📯
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgPJwv+JEf3TlslUjdcSz58oj7t8GSnSnNeIwcAh6zSPMC2xxFNDX7ISNolSOnM
Xw1ZFjx7XKxNsc1c/ZfCumK51H59fAXrDPGnXWST50W6lUMK7Ithr/AYtiviTbUL
xNeONWmAtjI4VjFRMtxyDJNJdkRdX7ex7P50av79lQ/DwKvIVcW4k4FPKHO183qr
FJKKPve6kCceMdufWaFHbHqTIUa1CtXOR1w3ZAfLH2X1SDxbSVTXwiJ3UlFwtpaI
jWF8Es9Jmhm3WEFTso+6Rzj61t0MNZa50CAdTV+gk/wptjgrG07ryqaHtCZLWJjf
LdZn0ZHiM5r0F+ArVzTGrtoqPZ2nWwRTaH5VxbK5UwfAe33woO2IFRdldlyJSaGN
cdlBEVYIZNxbgDehwJFytsnoDD+kGDb2bF9e202oJ6DQGP7uhb2vU8aUn7CYPa+X
gtA7UC41YvwVlFfURrOx7LYAFO6E/LChs3sJaZBdyhlrcETLTZxAhEphEHWqdMCZ
ecC5uhfq
=sVpp
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `fe1f05c2963a24a1253bb29b2e921091390831e2f2b6c031b2109dc1573157b6  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108fe1f05c2963a24a1253bb29b2e921091390831e2f2b6c031b2109dc1573157b6f0102835b3c81add50dfe77cf8dd50e29ea308fff01026c83022802cfeb308ec1e4dd145192a08f1045e41bcb6f0080ddcf616f2adbb6f0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0103015d844bd034c39dd1bba5fb521985308f1045e41bcb6f008c96fd7644e02c4de0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010b578451255a8f89af96624f0244ed03208f1045e41bcb6f008fcf833cfcd4da7740083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010c9d97ea8bb64d71d0c73e96c0876403908f1045e41bcb6f00843025273c3c6c3160083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

**Comment by jonatack** [comment]
ACK 0e0fa27 code review, built/ran tests/bitcoind

**Review by jonatack — COMMENTED** [review]
ACK 0e0fa27 code review, built/ran tests/bitcoind

**Comment by practicalswift** [comment]
ACK 0e0fa27acb74b4f0075afcf59a0dff51a21baddb -- diff looks correct

**Pull request merged**